Catch exceptions from UserObject evaluation

bwspenc opened the pull request:
closes #26685

bwspenc committed:
Catch exception during UserObject evaluation closes #26685

Also make a number of changes to make exception handling more
consistent:
1) Add a version of FEProblemBase::setException() that takes the curent
   exec flag as an argument, and sets an exception only if it's one
   that can be recovered from.
2) Add FEProblemBase::resetState() and call it from within
   checkExceptionAndStopSolve() if the solve will be retried.
3) Restructure many of the try/catch blocks to remove an unnecessary
   level of nesting.
4) Make the messages printed out for exceptions more consistent. It's
   a start, but there's more to be done here.
5) Make minor edits to the documentation for MooseException to clarify
   a few points.

bwspenc commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -6758,6 +6763,8 @@ FEProblemBase::computeBounds(NonlinearImplicitSystem & libmesh_dbg_var(sys),
   _lower.swap(lower);
   _upper.swap(upper);
 
+  //BWS: In all other cases where this is called, it's within a try/catch to strip
```
This is a note to self that I will remove when I have resolved this outstanding question.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2033783/) on 9873e59 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26686/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26686/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 796cb8aa68a15d61eb82eb353bb6582b01c6634f`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2033826/) on a555814 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26686/site)

This comment will be updated on new commits.

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -6758,6 +6779,8 @@ FEProblemBase::computeBounds(NonlinearImplicitSystem & libmesh_dbg_var(sys),
   _lower.swap(lower);
   _upper.swap(upper);
 
+  // BWS: In all other cases where this is called, it's within a try/catch to strip
+  // out the exception. Should that be done here too?
   checkExceptionAndStopSolve();
```
Yes we should

lindsayad replied:
Actually no I don't think we should. I am not 100% confident that this function is called during a solve, in which case I would rather have an uncaught exception that leads to simulation termination than a silenced exception

bwspenc replied:
OK, I'll just remove that comment and consider that resolved.

I do wonder whether it's worthwhile to add a parameter to `checkExceptionAndStopSolve()` (or another variant of it) that makes it not throw an exception. There are an awful lot of try/catch blocks around calls to this just to strip out the exception.

lindsayad replied:
Ok I looked into this more. This is called during setup of the nonlinear solver, so not during the Newton iterations. So if we have an exception at this point, then it is like having an exception for execution flags `!EXEC_LINEAR && !EXEC_NONLINEAR`, in which case we should mooseError. So I think in this function we should wrap the `_aux->compute` in a `try` and then in the `catch` we should `mooseError`

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -6297,10 +6332,6 @@ FEProblemBase::computeResidualAndJacobian(const NumericVector<Number> & soln,
     }
     catch (MooseException & e)
     {
-      _console << "\nA MooseException was raised during Auxiliary variable computation.\n"
-               << "The next solve will fail, the timestep will be reduced, and we will try again.\n"
-               << std::endl;
```
It looks like we would...

bwspenc replied:
I agree that this is valuable information that we don't want to lose. I believe what I was thinking was that this would get printed out later, but in looking at the output from the test case that motivated this (which also triggered some of these), I'm not getting what I wanted. I'll look into this.

bwspenc replied:
I'm pretty sure this is how we want this. The part about the solution failing and the step getting cut is printed out in `FEProblemBase::checkExceptionAndStopSolve()`

bwspenc commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -488,6 +501,13 @@ class FEProblemBase : public SubProblem, public Restartable
    */
   virtual void checkExceptionAndStopSolve(bool print_message = true);
 
+private:
+  /**
+   * Reset state of this object in preparation for the next evaluation.
+   */
+  virtual void resetState();
```
Happy to do that. I wasn't sure what the preference was.

bwspenc replied:
This is done

bwspenc commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -4304,216 +4304,218 @@ FEProblemBase::computeUserObjectsInternal(const ExecFlagType & type,
 
-  // iterate over execution order groups
-  for (const auto execution_group : execution_groups)
+  try
```
Makes sense, will do.

bwspenc replied:
Done

bwspenc commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5852,6 +5854,21 @@ FEProblemBase::setException(const std::string & message)
 }
 
+void
+FEProblemBase::setException(const ExecFlagType & type, const std::string & message)
```
Good catch! I had meant to use it on lines 4496 and 4502 of this file in place of the now-removed `process_possibly_handleable_exception`. I'll switch those over to use this.

bwspenc replied:
I'm calling this version of `setException` in most cases now.

bwspenc replied:
I ended up just getting rid of this new version of setException() and moved the additional checking elsewhere, so this isn't relevant anymore.

bwspenc commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -6554,7 +6575,7 @@ FEProblemBase::computeResidualTags(const std::set<TagID> & tags)
   _current_nl_sys->computeResidualTags(tags);
   _safe_access_tagged_vectors = true;
-  reset_state();
+  resetState();
```
This line gets hit during normal execution with no exceptions, which led me to believe that we need to do that resetting stuff all the time.

lindsayad replied:
of course, silly me

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5852,6 +5854,21 @@ FEProblemBase::setException(const std::string & message)
+void
+FEProblemBase::setException(const ExecFlagType & type, const std::string & message)
+{
+  if (type == EXEC_LINEAR || type == EXEC_NONLINEAR)
```
So thinking again about this, I do think we should limit this to `EXEC_LINEAR`. This is because our stopping mechanism in `NonlinearSystem::stopSolve()` is `SNESSetFunctionDomainError(solver.snes());` which is not going to be checked by PETSc during a nonlinear/Jacobian evaluation

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -4304,216 +4304,218 @@ FEProblemBase::computeUserObjectsInternal(const ExecFlagType & type,
+  }
+  catch (MooseException &)
+  {
+    // Just end now. We've inserted our NaN into the residual vector
```
Paired with my comment above, can we `mooseAssert(type == EXEC_LINEAR, "We should only be continuing if this is a linear/residual evaluation and consequently we know we've inserted a NaN in the residual vector");` or some message like that right here? Then we are confident in this comment

bwspenc committed:
Address review comments

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2070395/) on 37050b8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26686/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26686/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7cc4aedd3b71997307aaaf7cbae1810fd32698eb`

bwspenc commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5872,7 +5891,12 @@ FEProblemBase::checkExceptionAndStopSolve(bool print_message)
     if (_communicator.rank() == 0 && print_message)
-      Moose::err << _exception_message << std::endl;
+    {
+      _console << "\nThe following exception was caught in FEProblemBase:\n";
```
Note that this was going to `Moose::err` before, but that made the output weird. I think the standard console output is where we want this.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2072138/) on ad9f612 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26686/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26686/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7cc4aedd3b71997307aaaf7cbae1810fd32698eb`

bwspenc commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -6521,27 +6553,16 @@ FEProblemBase::computeResidualTags(const std::set<TagID> & tags)
   }
   catch (MooseException & e)
   {
-    _console << "\nA MooseException was raised during Auxiliary variable computation.\n"
-             << "The next solve will fail, the timestep will be reduced, and we will try again.\n"
-             << std::endl;
```
I think this is resolved per the discussion in the other comment.

bwspenc committed:
Move execution stage check, other misc cleanup

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2082403/) on 6cfb6a4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26686/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26686/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d6b47a87651987daded2fc2d7a29f1496d432dd4`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2082457/) on b93a877 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26686/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26686/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d6b47a87651987daded2fc2d7a29f1496d432dd4`

lindsayad commented:
@bwspenc just mention me here when this is ready for me to take over and investigate the automatic scaling issue

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2089053/) on 3d52cc5 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c8158d</th>
      <th colspan=4>#26686 3d52cc</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.29%</td>
      <td>85.28%</td>
      <td>-0.01%</td>
      <td>68.05%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>100806</td>
      <td>100812</td>
      <td>+6</td>
      <td>115</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17386</td>
      <td>17405</td>
      <td>+19</td>
      <td>54</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26686/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26686/coverage/framework)


## Modules coverage

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>c8158d</th>
      <th colspan=4>#26686 3d52cc</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.87%</td>
      <td>84.87%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>27122</td>
      <td>27122</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4834</td>
      <td>4834</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26686/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26686/coverage/solid_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26686/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 68.05% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Add a clearCurrentJacobianVectorTags method

Add a call to it in `resetState`

lindsayad committed:
Allow exception handling on EXEC_NONLINEAR

We will call `SNESSetJacobianDomainError` in this case

lindsayad committed:
Remove duplicate checkExceptionAndStopSolve in computeBounds

lindsayad committed:
Allow throwing exceptions while evaluating dampers

Dampers are computed during the `compute_postcheck` call-back from
PETSc. Right after calling the user-supplied postcheck method, PETSc
immediately does a call to `SNESComputeFunction` which will check
for the domain error that we set during the postcheck

lindsayad committed:
Check for exceptions after running ComputeJacobianForScalingThread

lindsayad committed:
Don't attempt to solve if automatic scaling calculation fails

This commit also includes some refactor meant to reduce code
duplication between `NonlinearSystem` and `NonlinearEigenSystem`

lindsayad committed:
Update solid mechanics AD return mapping test

When computing scaling once we now regard a failure on the initial
time step during the scaling calculation as not satisfying a computed
scaling criteria. This means that we keep attempting an auto-scaling
calculation, after cutting the time-step, until we get an exceptionless
scaling calculation. This is different from the past where even if
something bad happened during the scaling calculation we accepted
the partial results of the scaling calculation.

Differences in old and new scaling behavior led to a different solve
history for this test. So now I am just manually supplying the scaling
factors to reproduce the old behavior

bwspenc commented on framework/src/systems/NonlinearSystem.C:
```diff
@@ -234,29 +212,27 @@ NonlinearSystem::solve()
-    _Re_time->close();
-  _Re_non_time->close();
+
+  if (exec_flag == EXEC_LINEAR || exec_flag == EXEC_POSTCHECK)
```
I assume these should match what is in `FEProblemBase::checkExceptionAndStopSolve()`, right? Is it intentional that `EXEC_NONLINEAR` is missing?

lindsayad replied:
No. That's handled a few lines later

bwspenc replied:
Duhh... I see that now!

lindsayad committed:
Much more straightforward try-catch design

The design is to put the `try` at the top of each method. The `catch` should
dispatch to `handleException`. An outer `try-catch` can also be applied and
should be applied when we are in an outer method like
`computeResidualTags/computeJacobianTags/computeResidualAndJacobian`. When we
are in an "inner" but generic method like `computeUserObjectsInternal` there
should not be an outer `try-catch` because we must allow the exception coming
from `handleException` to propagate up higher to a place where we know whether
we can appropriately handle it and cut the timestep else error out

lindsayad commented:
Ok, I'm hoping after this most recent commit the only (real) failures are the two "high-Zr" tests in BISON. These tests throw exceptions on initial because the user supplies bad initial conditions. @bwspenc if you can decide what to do about those, that would be much appreciated

lindsayad committed:
Make cleaner exception messages

lindsayad committed:
Catch std::exception

lindsayad committed:
Remove extraneous semicolons after method definitions

lindsayad committed:
Add ability to regard general exceptions as errors

lindsayad committed:
Convert ERRMSG to CRASH for python unit test

lindsayad committed:
Always abort for PetscSolverExceptions

lindsayad commented:
@loganharbour @milljm I think that at some point a wasp submodule update must have been supplied in this PR. Now I'm stuck with a wasp on conda that is from this PR I think and that's why I'm getting these continued conda recipe failures. Is there any way to fix that and force the recipes to use the regular conda channel?

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2107108/) on 78f9c65 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2107122/) on 78f9c65 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 85.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 89.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

loganharbour commented:
> @loganharbour @milljm I think that at some point a wasp submodule update must have been supplied in this PR. Now I'm stuck with a wasp on conda that is from this PR I think and that's why I'm getting these continued conda recipe failures. Is there any way to fix that and force the recipes to use the regular conda channel?

Should be good now

lindsayad commented:
awesome, thank you!

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/2108981/) on a2c30d9 : invalidated by @bwspenc

Issue causing failing bison tests should be fixed by a recently-merged bison update

bwspenc commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5870,28 +5857,69 @@ FEProblemBase::checkExceptionAndStopSolve(bool print_message)
+      throw MooseException(_exception_message);
+    }
+    else
+      mooseError("The following parallel-communicated exception was detected during " +
```
@lindsayad What do you think about this message? Would you rather just call it a "MooseException" like we did before? Maybe a "parallel-communicated MooseException"? I kind of like including the parallel bit because it clarifies the fact that it has been communicated in parallel already.

lindsayad replied:
I like this message. Yes the parallel bit is good. But also I like that it *doesn't* say MooseException since we handle more exception types than just MooseExceptions

bwspenc replied:
OK, great -- let's leave it as is.

bwspenc commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -2816,11 +2816,20 @@ NonlinearSystemBase::computeJacobianInternal(const std::set<TagID> & tags)
       // they are almost always used in conjunction with Kernels
       ComputeJacobianForScalingThread cj(_fe_problem, tags);
       Threads::parallel_reduce(elem_range, cj);
+      // BWS TODO Add a call to FEProblemBase::checkExceptionAndStopSolve() here
```
@lindsayad I need to delete this before we merge this PR. You took care of that a few lines below. I want to see if the currently running tests pass before I go update this.

lindsayad replied:
K I removed it

lindsayad reviewed (approved):
approving Ben's work

lindsayad commented:
@bwspenc if you approve of what I've done, I think we can get this in

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/2111715/) on 3079d11 : invalidated by @bwspenc

bwspenc reviewed (commented):
Approving Alex's changes
